runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+		_ = conn.Close()
+		err = db.Close()
+		return err
+	}, nil
+}
+
+func (d *db) openDBAndAttach(ctx context.Context, read bool) (*sqlx.DB, error) {
+	// open the db
+	var (
+		dsn      *url.URL
+		err      error
+		settings map[string]string
+		path     string
+	)
+	if read {
+		dsn, err = url.Parse("") // in-memory
+		settings = d.opts.ReadSettings
+		path = d.readPath
+	} else {
+		path = d.writePath
+		dsn, err = url.Parse(filepath.Join(path, "stage.db"))
+		settings = d.opts.WriteSettings
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	query := dsn.Query()
+	for k, v := range settings {
+		query.Set(k, v)
+	}
+	dsn.RawQuery = query.Encode()
+	connector, err := duckdb.NewConnector(dsn.String(), func(execer driver.ExecerContext) error {
+		for _, qry := range d.opts.InitQueries {
+			_, err := execer.ExecContext(context.Background(), qry, nil)
+			if err != nil && strings.Contains(err.Error(), "Failed to download extension") {
+				// Retry using another mirror. Based on: https://github.com/duckdb/duckdb/issues/9378
+				_, err = execer.ExecContext(context.Background(), qry+" FROM 'http://nightly-extensions.duckdb.org'", nil)
+			}
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	db := sqlx.NewDb(otelsql.OpenDB(connector), "duckdb")
+
+	err = otelsql.RegisterDBStatsMetrics(db.DB, otelsql.WithAttributes(attribute.String("db.system", "duckdb"), attribute.String("db_identifier", d.dbIdentifier)))
+	if err != nil {
+		return nil, fmt.Errorf("registering db stats metrics: %w", err)
+	}
+
+	err = db.PingContext(ctx)
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	err = d.attachDBs(ctx, db, path, read)
+	if err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	// 2023-12-11: Hail mary for solving this issue: https://github.com/duckdblabs/rilldata/issues/6.
+	// Forces DuckDB to create catalog entries for the information schema up front (they are normally created lazily).
+	// Can be removed if the issue persists.
+	_, err = db.ExecContext(context.Background(), `
+		select
+			coalesce(t.table_catalog, current_database()) as "database",
+			t.table_schema as "schema",
+			t.table_name as "name",
+			t.table_type as "type", 
+			array_agg(c.column_name order by c.ordinal_position) as "column_names",
+			array_agg(c.data_type order by c.ordinal_position) as "column_types",
+			array_agg(c.is_nullable = 'YES' order by c.ordinal_position) as "column_nullable"
+		from information_schema.tables t
+		join information_schema.columns c on t.table_schema = c.table_schema and t.table_name = c.table_name
+		group by 1, 2, 3, 4
+		order by 1, 2, 3, 4
+	`)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+func (d *db) attachDBs(ctx context.Context, db *sqlx.DB, path string, read bool) error {
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	var views []string
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
```

Comment:
Not sure, but feels like the table traversal logic shows up a few places – did you consider a generic walker function that calls a callback for each table version, e.g. `d.walkTables(ctx, func(name, path, version string) error) error`?

Reply:
It is at two places(syncRead and syncWrite). Both have subtle variations and not sure if `walkTables` will help. I think it will anyways be removed if we go ahead with non duplicated write and read tables.